firedrake/interpolation.py

```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
-        self.arguments = arguments
-        self.nargs = len(arguments)
-        self.freeze_expr = freeze_expr
+
+    def __new__(cls, expr, V, **kwargs):
+        target_mesh = V.ufl_domain()
+        source_mesh = extract_unique_domain(expr) or target_mesh
+        if target_mesh is not source_mesh:
+            if isinstance(target_mesh.topology, firedrake.mesh.VertexOnlyMeshTopology):
+                return object.__new__(SameMeshInterpolator)
+            else:
+                return object.__new__(CrossMeshInterpolator)
+        else:
+            return object.__new__(SameMeshInterpolator)
+
+    def __init__(
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_points=False,
+    ):
         self.expr = expr
         self.V = V
+        self.subset = subset
+        self.freeze_expr = freeze_expr
+        self.access = access
         self.bcs = bcs
+        self._allow_missing_points = allow_missing_points
+        self.callable = None
 
-    @PETSc.Log.EventDecorator()
-    @annotate_interpolate
-    def interpolate(self, *function, output=None, transpose=False):
-        """Compute the interpolation.
+    def interpolate(self, *args, **kwargs):
+        """
+        Compute the interpolation.
 
         :arg function: If the expression being interpolated contains an
             :class:`ufl.Argument`, then the :class:`.Function` value to
             interpolate.
         :kwarg output: Optional. A :class:`.Function` to contain the output.
         :kwarg transpose: Set to true to apply the transpose (adjoint) of the
               interpolation operator.
+        :kwarg default_missing_val: For interpolation across meshes: the
+            optional value to assign to DOFs in the target mesh that are
+            outside the source mesh. If this is not set and an ``output``
+            :class:`.Function` is specified, then such DOF values are left
+            unchanged. If this is not set and no ``output`` :class:`.Function`
+            is specified, then the default value is zero. This does not effect
+            transpose interpolation.
+
         :returns: The resulting interpolated :class:`.Function`.
         """
+        return self.interpolate(*args, **kwargs)
+
+
+class CrossMeshInterpolator(Interpolator):
+    """
+    Interpolate a function from one mesh and function space to another.
+
+    Only works when the function space we interpolate into has point evaluation
+    nodes. A more general version would require finding the global coordinates
+    of all quadrature points of the target function space in the source mesh.
+
+    For arguments, see :class:`.Interpolator`.
+    """
+
+    def __init__(
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_points=False,
+    ):
+        if subset:
```

Comment:
It should definitely be documented that these options will not work. The only way a user would find out that `subset` etc won't work is by trying it and having it break.